codex/blockexchange/engine/engine.nim

```diff
@@ -427,10 +425,8 @@ proc wantListHandler*(
           address: e.address,
           `type`: BlockPresenceType.Have,
           price: price))
-      elif e.wantType == WantType.WantBlock:
```

Comment:
There is a good reason why I separated `WantBlock` into it's own branch. We used to schedule a task for every single `WantList` update, but obviously this isn't necessary and worst, it was severely affecting performance. We should not attempt to schedule tasks on `WantHave`, since this is just a presence update.

Reply:
> We track all peerWants instead of only the wantBlocks. This is according to spec, and solves an issue in the case where a block is received/detected after the want-list message.

For more context, this isn't actually true, we do track `WantHave` messages, but we respond to them immediately with a presence message.

Reply:
OK, I see the problem with the scheduling of tasks for every wantList. Makes perfect sense and I'll fix that right now.

Regarding the tracking of wantHaves: Responding immediately is good but not enough. We're expected to keep the want in the peerCtx. If we encounter the block later, we should still react to the want with a presence message. The tests in `NetworkStore - multiple nodes` in testblockexc.nim actually check this behavior perfectly. They were just previously passing because 1) we were sending wantBlocks when we shouldn't and 2) we were keeping the wantBlocks only. So fixing the send of wantBlock caused this correct test to reveal the other issue.

I would like to take a closer look at exactly how the stored wants will trigger messages when blocks are received. But I'm trying to keep the scope of this PR small.

Reply:
I've ran some more tests after applying these changes. The performance improvement is noticeable. I think reliability is also up, but that takes a bit longer to be sure about.

Reply:
OK, I see. Then, what we need to do is to add these to the `BlockExcPeerCtx.peerWants` list in addition to responding to the request.

---------------

For a later PR, we should take a closer look at how these lists are kept, for example, the `peerWants` is a simple `seq`, which doesn't have any sort of caps, nor ordering. 

Some lists, like the `peerWants`, in addition to be capped, should also allow for ordering, for example `WantListEntry` has a `priority` field, which should be respected.

Reply:
So, what actually needs to happen is that when adding a `DontHave` to the `presence` seq, we also add it to `peerCtx.peerWants` list. We only need to remember the `DontHave`s, because the remote should keep track of the `Have`s we send back.

Reply:
Yes: Upgrading the peerWants to a capped, prioritized list is definitely needed.

"what needs to happen is that when adding a DontHave to the presence seq, we also add it to peerCtx.peerWants list."
In the PR, yes we are adding it to the peerWants regardless of Have/DontHave

"We only need to remember the DontHaves, because the remote should keep track of the Haves we send back."
I understand this to mean:
- If we Have, don't add the want to the peerCtx, only send presence.
- If we dont Have, add want to peerCtx, and send dontHave presence if asked. (e.sendDontHave)
For the record, e.sendDontHave is has always been false up to this point except for some tests and that's a good thing because the peerContext peerHave proc doesn't consider the have-flag currently.

Reply:
> If we Have, don't add the want to the peerCtx, only send presence.
> If we dont Have, add want to peerCtx, and send dontHave presence if asked. (e.sendDontHave)

Yep, thats what I mean.